src/auth/index.ts

```diff
@@ -28,6 +29,48 @@ export class Authenticator {
     this.options = options;
   }
 
+  private getCacheKeys() {
+    let clientId, secret;
+    if (this.isUsernamePassword()) {
+      const auth = this.options.auth as UsernamePasswordAuth;
+      [clientId, secret] = [auth.username, auth.password];
+    } else if (this.isServiceAccount()) {
+      const auth = this.options.auth as ServiceAccountAuth;
+      [clientId, secret] = [auth.client_id, auth.client_secret];
+    }
+    return { clientId, secret };
+  }
+
+  private getCache() {
+    return this.options.useCache ?? true ? inMemoryCache : noneCache;
+  }
+
+  clearCache() {
+    const { clientId, secret } = this.getCacheKeys();
+    if (!clientId || !secret) {
```

Comment:
I'd try to create generic function with callback that implements this strategy. This will reduce the code duplication

Reply:
I won't be able to use callback since I need to return the actuall value sometimes. Yet, I'll try to simplify the code